src/sample_inverted_index.cpp

```diff
@@ -0,0 +1,72 @@
+#include <string>
+#include <random>
+#include <fstream>
+
+#include "invert.hpp"
+
+#include "binary_freq_collection.hpp"
+
+#include "util/progress.hpp"
+
+#include "CLI/CLI.hpp"
+
+
+int main(int argc, char **argv) {
+
+    using namespace pisa;
+    std::string input_basename;
+    std::string output_basename;
+    float rate;
+
+    CLI::App app{"sample_inverted_index - a tool for sampling an inverted index."};
+    app.add_option("-c,--collection", input_basename, "Input collection basename")->required();
+    app.add_option("-o,--output", output_basename, "Output collection basename")->required();
+    app.add_option("-r,--rate", rate, "Sampling rate")->required();
+    CLI11_PARSE(app, argc, argv);
+
+    if(rate < 0 or rate > 1)
+    {
+        spdlog::error("Sampling rate should be between 0 and 1.");
+        std::abort();
+    }
+    binary_freq_collection input(input_basename.c_str());
+
+    // Copy '.sizes' file
+    std::ifstream  sis(input_basename + ".sizes");
+    std::ofstream  sos(output_basename + ".sizes");
+    sos << sis.rdbuf();
+    sis.close();
+    sos.close();
+
+    std::ofstream dos(output_basename + ".docs");
+    std::ofstream fos(output_basename + ".freqs");
+
+    auto document_count = static_cast<uint32_t>(input.num_docs());
+    write_sequence(dos, gsl::make_span<uint32_t const>(&document_count, 1));
+
+    {
+        pisa::progress progress("Sampling inverted index", input.size());
+        for (auto const &plist : input) {
+            std::vector<uint32_t> docs(plist.docs.begin(), plist.docs.end());
+            std::vector<uint32_t> freqs(plist.freqs.begin(), plist.freqs.end());
+
+            auto seed = unsigned ( std::time(0) );
+            std::mt19937 eng1(seed);
+            auto eng2 = eng1;
+            std::shuffle(docs.begin(), docs.end(), eng1);
+            std::shuffle(freqs.begin(), freqs.end(), eng2);
```

Comment:
This is actually incorrect. You're sampling documents and frequencies independently, so they don't match.

Moreover, there is a `std::sample` function that will do the sampling for you without having to copy and shuffle the entire lists. Plus, it is stable, so there's no need to sort after.

Here's a quick example of how it would work: https://gcc.godbolt.org/z/GAgeFP

Reply:
It should be correct as I am shuffling using the same seed and state. This is why I have eng1 and eng2. It might not be the cleanest, but it should work

Reply:
Oh, that's why. Again, take a look at the link, we can avoid copying all data and sorting.

Reply:
To be more precise, we still have to copy indices, but it's fewer copies and no sorting necessary.

Reply:
I decided to go with `std::sample` rather than more complicated solutions since it is not critical